vnext/Microsoft.ReactNative/TurboModulesProvider.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "TurboModulesProvider.h"
+
+using namespace winrt;
+using namespace Windows::Foundation;
+
+namespace winrt::Microsoft::ReactNative {
+/*-------------------------------------------------------------------------------
+  TurboModuleImpl
+-------------------------------------------------------------------------------*/
+
+class TurboModuleImpl : public facebook::react::TurboModule {
+ public:
+  TurboModuleImpl(
+      const std::string &name,
+      std::shared_ptr<facebook::react::CallInvoker> jsInvoker,
+      ReactModuleProvider reactModuleProvider)
+      : facebook::react::TurboModule(name, jsInvoker), m_reactModuleProvider(reactModuleProvider) {}
+
+  facebook::jsi::Value get(facebook::jsi::Runtime &runtime, const facebook::jsi::PropNameID &propName) override {
+    VerifyElseCrash(false);
+  }
+
+ private:
+  ReactModuleProvider m_reactModuleProvider;
+};
+
+/*-------------------------------------------------------------------------------
+  TurboModulesProvider
+-------------------------------------------------------------------------------*/
+std::shared_ptr<facebook::react::TurboModule> TurboModulesProvider::getModule(
+    const std::string &moduleName,
+    const std::shared_ptr<facebook::react::CallInvoker> &callInvoker) noexcept {
+  auto it = m_moduleProviders.find(moduleName);
+  if (it == m_moduleProviders.end()) {
+    return nullptr;
+  }
+
+  return std::make_shared<TurboModuleImpl>(moduleName, callInvoker, it->second);
+}
+
+std::vector<std::string> TurboModulesProvider::getEagerInitModuleNames() noexcept {
+  return {};
+}
+
+void TurboModulesProvider::AddModuleProvider(
```

Comment:
Not sure what the android/ios semantics are here, but should we report an error if duplicate named modules are registered?